*** InstanceScript.cpp.old	2011-12-10 23:49:44.231550000 +0100
--- InstanceScript.cpp	2011-12-06 20:55:48.395550000 +0100
***************
*** 320,325 ****
--- 320,343 ----
      }
  }
  
+ // Complete Achievement for all players in instance
+ void InstanceScript::DoCompleteAchievement(uint32 achievement)
+ {
+     AchievementEntry const* pAE = GetAchievementStore()->LookupEntry(achievement);
+     Map::PlayerList const &PlayerList = instance->GetPlayers();
+ 
+     if (!pAE)
+     {
+         sLog->outError("TSCR: DoCompleteAchievement called for not existing achievement %u", achievement);
+         return;
+     }
+ 
+     if (!PlayerList.isEmpty())
+         for (Map::PlayerList::const_iterator i = PlayerList.begin(); i != PlayerList.end(); ++i)
+             if (Player *pPlayer = i->getSource())
+                 pPlayer->CompletedAchievement(pAE);
+ }
+ 
  // Update Achievement Criteria for all players in instance
  void InstanceScript::DoUpdateAchievementCriteria(AchievementCriteriaTypes type, uint32 miscValue1 /*= 0*/, uint32 miscValue2 /*= 0*/, Unit* unit /*= NULL*/)
  {
