*** Object.h.old	2011-12-10 23:49:43.922550000 +0100
--- Object.h	2011-12-06 20:55:48.032550000 +0100
***************
*** 781,786 ****
--- 781,788 ----
  
          Creature*   FindNearestCreature(uint32 entry, float range, bool alive = true) const;
          GameObject* FindNearestGameObject(uint32 entry, float range) const;
+         Player*     FindNearestPlayer(float range, bool alive = true);
+         std::list<Player*>  GetNearestPlayersList(float range, bool alive = true);
  
          void GetGameObjectListWithEntryInGrid(std::list<GameObject*>& lList, uint32 uiEntry, float fMaxSearchRange) const;
          void GetCreatureListWithEntryInGrid(std::list<Creature*>& lList, uint32 uiEntry, float fMaxSearchRange) const;
