*** Player.h.old	2011-12-10 23:49:44.002550000 +0100
--- Player.h	2011-12-06 20:55:48.089550000 +0100
***************
*** 2473,2478 ****
--- 2519,2525 ----
          AchievementMgr const& GetAchievementMgr() const { return m_achievementMgr; }
          void UpdateAchievementCriteria(AchievementCriteriaTypes type, uint32 miscValue1 = 0, uint32 miscValue2 = 0, Unit* unit = NULL);
          void CompletedAchievement(AchievementEntry const* entry);
+         bool HasAchieved(uint32 entry);
  
          bool HasTitle(uint32 bitIndex);
          bool HasTitle(CharTitlesEntry const* title) { return HasTitle(title->bit_index); }