*** ScriptedCreature.h.old	2011-10-10 17:28:18.773600000 +0200
--- ScriptedCreature.h	2011-12-06 20:55:47.249550000 +0100
***************
*** 24,29 ****
--- 24,30 ----
  #include "CreatureAIImpl.h"
  #include "InstanceScript.h"
  
+ #define MAX_AGGRO_PULSE_TIMER            5000
  #define CAST_PLR(a)     (dynamic_cast<Player*>(a))
  #define CAST_CRE(a)     (dynamic_cast<Creature*>(a))
  #define CAST_AI(a, b)   (dynamic_cast<a*>(b))
***************
*** 233,238 ****
--- 234,245 ----
          return heroic25;
      }
  
+     void SetImmuneToPushPullEffects(bool set)
+     {
+         me->ApplySpellImmune(0, IMMUNITY_EFFECT, SPELL_EFFECT_KNOCK_BACK, set);
+         me->ApplySpellImmune(0, IMMUNITY_MECHANIC, MECHANIC_GRIP, set);
+     }
+ 
      private:
          Difficulty _difficulty;
          uint32 _evadeCheckCooldown;
***************
*** 242,248 ****
  
  struct Scripted_NoMovementAI : public ScriptedAI
  {
!     Scripted_NoMovementAI(Creature* creature) : ScriptedAI(creature) {}
      virtual ~Scripted_NoMovementAI() {}
  
      //Called at each attack of me by any victim
--- 249,258 ----
  
  struct Scripted_NoMovementAI : public ScriptedAI
  {
!     Scripted_NoMovementAI(Creature* creature) : ScriptedAI(creature) 
! 	{
!         SetImmuneToPushPullEffects(true);
! 	}
      virtual ~Scripted_NoMovementAI() {}
  
      //Called at each attack of me by any victim
***************
*** 256,261 ****
--- 266,272 ----
          virtual ~BossAI() {}
  
          InstanceScript* const instance;
+         uint32 inFightAggroCheck_Timer;
          BossBoundaryMap const* GetBoundary() const { return _boundary; }
  
          void JustSummoned(Creature* summon);
***************
*** 279,284 ****
--- 290,296 ----
          void _EnterCombat();
          void _JustDied();
          void _JustReachedHome() { me->setActive(false); }
+         void _DoAggroPulse(const uint32 diff);
  
          bool CheckInRoom()
          {
